pep-9999.txt

```diff
@@ -0,0 +1,171 @@
+Version: $Revision$
+Last-Modified: $Date$
+Author: Kushal Das <mail@kushaldas.in>, Steve Dower <steve.dower@python.org>,
+        Donald Stufft <donald@stufft.io>, Nick Coghlan <ncoghlan@gmail.com>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 16-May-2018
+Python-Version: 3.8
+
+
+
+Abstract
+========
+
+This PEP proposes to add to Python a mechanism to have a ``__pypackages__``
+directory  per project to have its own independent set of installed packages.
+This will share the standard library with the base installed Python. This will
+avoid the steps to create or activate or deactivate "virtual environments". Any
+Python scripts will use the ``__pypackages__`` from the base directory of the 
+script.
+
+
+
+Motivation
+==========
+
+Python virtual environments have become an essential part of development and
+teaching workflow in the community, but at the same time, they create a
+barrier to entry for many folks. The following are few examples of the issues
+people get into while being introduced to Python (or programming for the first
+time).
+
+- How virtual environments work is a lot of information for anyone new. It takes
+  a lot of extra time and effort to explain them.
+
+- Different platforms and shell environments require different sets of commands
+  to work on the virtual environments. Any workshop or teaching environment with
+  people coming with different operating systems installed on the laptops create a
+  lot of confusion among the participants.
+
+- Virtual environments need to be activated on each opened terminal, if someone
+  creates/opens a new terminal, that by default does not get the same environment
+  as in a previous terminal with virtual environment activated.
+
+
+Specification
+=============
+
+When the Python binary is executed, it attempts to determine its
+prefix (which it stores in ``sys.prefix``), which is then used to find
+the standard library and other key files, and by the ``site`` module
+to determine the location of the site-package directories.  Currently
+the prefix is found (assuming ``PYTHONHOME`` is not set) by first
+walking up the filesystem tree looking for a marker file (``os.py``)
+that signifies the presence of the standard library, and if none is
+found, falling back to the build-time prefix hard coded in the binary.
+
+This PEP proposes to add a new step in this process. If a ``__pypackages__``
+directory is found in the current working directory, then it will be included in
+``sys.path`` after the current working directory and just before the system
+site-packages. This way, if the Python executable starts in the given project
+directory, it will automatically find all the dependencies inside of
+``__pypackages__``.
+
+In case Python scripts, Python will try to find ``__pypackages__`` in the
+base directory of the script. If found, then it will be used, otherwise the
+Python will behave as it does currently.
```

Comment:
Given the feedback we've heard so far, I think we should describe the `python myscript.py` case first, since people have been nervous about using the CWD for this. After describing the script case, we can say "we don't have a script directory, so use CWD, and if we then have `-m` then we'll use `__pypackages__` as well"

Reply:
From talking with kushal... I think that the -m case is the only one that uses CWD... So perhaps I'd phrase it as being relative to the script for all of the first ones.  And then the last one would say that "with -m, which has to search for the script to use rather than having a script specified, we search for __pypackages__ in the CWD"  (wording needs adjustment).

Reply:
There's also the interactive case, where we'd want to be able to `import` packages that are installed in `__pypackages__`. But I think the semantics of `-m` and an interactive `import` should be the same anyway.